analyzers/tests/SonarAnalyzer.UnitTest/Rules/Utilities/TokenTypeAnalyzerTest.Classifier.cs

```diff
@@ -222,4 +222,677 @@ public class Test {
+    // We cannot be sure without calling the model but we assume this will rarely be a type
+    [DataRow("class TypeParameter: System.Collections.Generic.[t:List]<System.[t:Exception]> { }", false)]
+    [DataRow("class TypeParameter: [u:System].[u:Collections].[u:Generic].List<[u:System].Exception> { }", false)]
+    [DataRow("class TypeParameter: [t:List]<[t:List]<[t:Exception]>> { }", false)]
+    [DataRow("class TypeParameter: [t:Outer].[t:Inner] { }", false)] // To decide what we want to do
+    [DataRow("class TypeParameter: [t:HashSet]<[t:List]<[k:int]>> { }", false)]
+    [DataRow("class TypeParameter<T> : [t:List]<T> where T: [t:List]<[t:Exception]> { }", false)]
+    [DataRow("class TypeParameter<[t:T]> : List<[t:T]> where [t:T]: List<Exception> { }", false)]
+    [DataRow("class TypeParameter<[t:T1], [t:T2]> where [t:T1] : class where [t:T2] : [t:T1], new() { }", false)]
+    public void IdentifierToken_TypeParameters(string syntax, bool allowSemanticModel = true) =>
+        ClassifierTestHarness.AssertTokenTypes($$"""
+            using System;
+            using System.Collections.Generic;
+            class Outer { public class Inner { } }
+            {{syntax}}
+            """ /*, allowSemanticModel */);
+
+    [DataTestMethod]
+    [DataRow("class [t:BaseTypeList]: System.[t:Exception] { }", false)]
+    [DataRow("class BaseTypeList: [u:System].Exception { }", false)]
+    [DataRow("class BaseTypeList: [t:Outer].[t:Inner] { }", false)]
+    [DataRow("""
+             class BaseTypeList: [t:IFormattable], [t:IFormatProvider]
+             {
+                public object? GetFormat(Type? formatType) => default;
+                public string ToString(string? format, IFormatProvider? formatProvider) => default;
+             }
+             """, false)]
+    public void IdentifierToken_BaseTypeList(string syntax, bool allowSemanticModel = true) =>
+        ClassifierTestHarness.AssertTokenTypes($$"""
+            using System;
+            class Outer { public class Inner { } }
+            {{syntax}}
+            """ /*, allowSemanticModel */);
+
+    [DataTestMethod]
+    [DataRow("class", false)]
+    [DataRow("struct", false)]
+    [DataRow("record", false)]
+    [DataRow("record struct", false)]
+#if NET
+    [DataRow("interface", false)]
+#endif
+    public void IdentifierToken_BaseTypeList_DifferentTypeKind(string syntax, bool allowSemanticModel = true) =>
+        ClassifierTestHarness.AssertTokenTypes($$"""
+            {{syntax}} X : [u:System].[t:IFormattable] { public string ToString(string? format, System.IFormatProvider? formatProvider) => null; }
+            """ /*, allowSemanticModel */);
+
+    [DataTestMethod]
+    [DataRow("_ = typeof([u:System].[t:Exception]);", false)]
+    [DataRow("_ = typeof([u:System].[u:Collections].[u:Generic].[t:Dictionary]<,>);", false)]
+    [DataRow("_ = typeof([t:Inner]);", false)]
+    [DataRow("_ = typeof([t:C].[t:Inner]);", false)]
```

Comment:
If we want to get the type of `C` we will most probably need a semantic model call.